src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/ChangeTracker.cs

```diff
@@ -0,0 +1,29 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace RadixDlt.NetworkGateway.PostgresIntegration.LedgerExtension;
+
+internal class ChangeTracker<TKey, TValue>
+    where TKey : notnull
+{
+    private readonly Dictionary<TKey, TValue> _store = new();
+    private readonly List<TKey> _insertionOrder = new();
+
+    public IList<TKey> Keys => _insertionOrder;
+
+    public TValue GetOrAdd(TKey key, Func<TKey, TValue> factory)
+    {
+        return _store.GetOrAdd(key, _ =>
+        {
+            _insertionOrder.Add(key);
+
+            return factory(key);
+        });
+    }
+
+    public IEnumerable<KeyValuePair<TKey, TValue>> AsEnumerable()
```

Comment:
Do we need to use enumrable? I'm not big fan of that interface, can we use some other interace i.e IReadOnlyCollection or some concrete implementation?

Reply:
Enumerable is a right choice here as it does **not** result in needless memory allocation. Just not being fan of something isn't enough here in my opinion. ;-)

Reply:
Ah, and I forgot about the most important one. This interface protects us against concurrent modifications. An attempt to `.GetOrAdd` while iterating will result in regular `Collection was modified; enumeration operation may not execute.` exception just as with regular Dictionary. This is important.

Reply:
well I don't see any place where we modify that so we're safe with using `IReadOnlyCollection`. and i'm not a fan of `IEnumerable` because it leads into potentially hard to detect errors when it's hard to track when and where collection has changed. In that example it's not an issue but i'd like to avoid that interface if possible. Feel free to ignore that comment if you strongly disagree.

Reply:
> well I don't see any place where we modify that so we're safe with using IReadOnlyCollection.
The idea behind the code is to make sure it will throw the moment we start doing something incorrectly. 

> and i'm not a fan of IEnumerable because it leads into potentially hard to detect errors [...]
Concurrent dictionary modifications during iteration are even more difficult to detect/spot/fix. 

> Feel free to ignore that comment if you strongly disagree.
I have a rather strong opinion this time. I'm going to ignore the comment then.